### PR TITLE
[enhancement](compress) reuse compression ctx and buffer

### PR DESCRIPTION
**Pull request by yixiutt**
Reuse compression ctx and buffer.
Use a global instance for every compression algorithm, and use a thread saft buffer pool to reuse compression buffer, pool size is equal to max parallel thread num in compression, and this will not be too large.

Test shows this feature increase 5% of data import and compaction.

# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/util/block_compression.cpp [hunk 1]
```diff
@@ -368,76 +555,87 @@ class ZlibBlockCompression : public BlockCompressionCodec {
...
     }
 
-    size_t max_compressed_len(size_t len) const override {
+    size_t max_compressed_len(size_t len) override {
         // one-time overhead of six bytes for the entire stream plus five bytes per 16 KB block
         return len + 6 + 5 * ((len >> 14) + 1);
     }
+
+    bool exceed_max_compress_len(size_t uncompressed_size) override {
+        (void)uncompressed_size;
```
[the remark this answers is not in the document]
Comment (yixiutt) [reply be/src/util/block_compression.cpp]: useless and deleted

File: be/src/util/block_compression.cpp [hunk 2]
```diff
@@ -586,74 +879,56 @@ class GzipBlockCompression final : public ZlibBlockCompression {
...
 
 Status get_block_compression_codec(segment_v2::CompressionTypePB type,
-                                   std::unique_ptr<BlockCompressionCodec>& codec) {
-    BlockCompressionCodec* ptr = nullptr;
+                                   BlockCompressionCodec** codec) {
     switch (type) {
     case segment_v2::CompressionTypePB::NO_COMPRESSION:
-        codec.reset(nullptr);
+        *codec = nullptr;
         return Status::OK();
```
Comment (HappenLee) [inline be/src/util/block_compression.cpp:886]: seems should break like other ？

File: be/src/util/block_compression.cpp [hunk 3]
```diff
@@ -586,74 +879,56 @@ class GzipBlockCompression final : public ZlibBlockCompression {
...
 
 Status get_block_compression_codec(tparquet::CompressionCodec::type parquet_codec,
-                                   std::unique_ptr<BlockCompressionCodec>& codec) {
-    BlockCompressionCodec* ptr = nullptr;
+                                   BlockCompressionCodec** codec) {
     switch (parquet_codec) {
     case tparquet::CompressionCodec::UNCOMPRESSED:
-        codec.reset(nullptr);
+        *codec = nullptr;
         return Status::OK();
```
Comment (HappenLee) [inline be/src/util/block_compression.cpp:914]: seems should break like other ？

Comment (yixiutt) [reply be/src/util/block_compression.cpp:914]: fixed

File: be/src/vec/core/block.cpp [hunk 4]
```diff
@@ -689,15 +690,16 @@ Status Block::filter_block(Block* block, int filter_column_id, int column_to_kee
     return Status::OK();
 }
 
-Status Block::serialize(PBlock* pblock, size_t* uncompressed_bytes, size_t* compressed_bytes,
+/*Status Block::serialize(PBlock* pblock, size_t* uncompressed_bytes, size_t* compressed_bytes,
```
Comment (BiteTheDDDDt) [inline be/src/vec/core/block.cpp:693]: Maybe we can delete those code directly.

File: be/src/vec/core/block.cpp [hunk 5]
```diff
@@ -711,10 +713,10 @@ Status Block::serialize(PBlock* pblock, std::string* compressed_buffer, size_t*
 
     // serialize data values
     // when data type is HLL, content_uncompressed_size maybe larger than real size.
-    std::string* column_values = nullptr;
+    std::string column_values;
     try {
-        column_values = pblock->mutable_column_values();
-        column_values->resize(content_uncompressed_size);
+        //column_values = pblock->mutable_column_values();
```
Comment (BiteTheDDDDt) [inline be/src/vec/core/block.cpp:718]: ditto

File: be/src/util/block_compression.cpp [hunk 6]
```diff
@@ -28,47 +28,92 @@
...
+        } else {
+            // reuse context buffer if max_len < MAX_COMPRESSION_BUFFER_FOR_REUSE
+            context->buffer.resize(max_len);
+            compressed_buf.data = reinterpret_cast<char*>(context->buffer.data());
+            compressed_buf.size = max_len;
         }
-        auto compressed_len =
-                LZ4_compress_default(input.data, output->data, input.size, output->size);
+
+        int32_t acc = 1;
```
Comment (BiteTheDDDDt) [inline be/src/util/block_compression.cpp:100]: Does we should definal acc as a constexpr static variable?

File: be/src/util/block_compression.cpp [hunk 7]
```diff
@@ -78,122 +123,260 @@ class Lz4BlockCompression : public BlockCompressionCodec {
...
+            context->ctx = LZ4_createStream();
+            if (context->ctx == nullptr) {
+                delete context;
+                return Status::InvalidArgument("LZ4_createStream error");
+            }
+            *out = context;
+            return Status::OK();
         }
-        return LZ4_compressBound(len);
+        *out = _ctx_pool.back();
```
Comment (BiteTheDDDDt) [inline be/src/util/block_compression.cpp:145]: Does we can move lock to this line?

File: be/src/vec/core/block.cpp (hunk 4 above)
Comment (yixiutt) [reply be/src/vec/core/block.cpp]: done

File: be/src/util/block_compression.cpp (hunk 7 above)
Comment (yixiutt) [reply be/src/util/block_compression.cpp:144]: done

File: be/src/vec/core/block.cpp (hunk 5 above)
Comment (yixiutt) [reply be/src/vec/core/block.cpp]: done

**Review by BiteTheDDDDt — APPROVED** [review]
LGTM

File: be/src/util/block_compression.h [hunk 8]
```diff
@@ -44,20 +48,23 @@ class BlockCompressionCodec {
...
     // for decompressed data.
     // Size of decompressed data will be set in output's size.
-    virtual Status decompress(const Slice& input, Slice* output) const = 0;
+    virtual Status decompress(const Slice& input, Slice* output) = 0;
 
     // Returns an upper bound on the max compressed length.
-    virtual size_t max_compressed_len(size_t len) const = 0;
+    virtual size_t max_compressed_len(size_t len) = 0;
+
+    virtual bool exceed_max_compress_len(size_t uncompressed_size);
```
[the remark this answers is not in the document]
Comment (yixiutt) [reply be/src/util/block_compression.h:67]: actually lz4 should return LZ4_MAX_INPUT_SIZE but i just copy the logic in previous version,so i left a virtual here

**Pull request merged**